# scrabbitR

An R package containing utility functions used as part of the Ton et al. 2022<sup>1</sup> paper exploring rabbit development with single-cell transcriptomics. 


> **Rabbit Development as a Model for Single Cell Comparative Genomics**

>
> Mai-Linh Ton<sup>1,2,\*</sup> Daniel Keitley<sup>3,*</sup>, Bart Theeuwes<sup>2</sup>, Carolina Guibentif<sup>4</sup>, Jonas Ahnfelt-Rønne<sup>5</sup>, Thomas Kjærgaard Andreassen<sup>5</sup>, Fernando J. Calero-Nieto<sup>2</sup>, Ivan Imaz-Rosshandler<sup>6</sup>, Blanca Pijuan-Sala<sup>7</sup>, Jennifer Nichols<sup>2</sup>, Èlia Benito-Gutiérrez<sup>3</sup>, John Marioni<sup>7,8,9</sup>, Berthold Göttgens<sup>1,2</sup>

See project webpage [here](https://marionilab.github.io/RabbitGastrulation2022/) for more info. 



## Installation

We recommend installing within a [conda environment](https://docs.conda.io/projects/conda/en/latest/index.html). The environment originally used in our analysis can be found [here](https://github.com/MarioniLab/RabbitGastrulation2022/blob/master/envs/scrabbitr_env.yml).

```R
# Inside R session
devtools::install_github("https://github.com/dkeitley/scrabbitr")
```


## Example usage
Comparing neighbourhoods across rabbit and mouse - [link](https://github.com/MarioniLab/RabbitGastrulation2022/blob/master/notebooks/compare_nhoods.ipynb)

Predicting cell types across species - [link](https://github.com/MarioniLab/RabbitGastrulation2022/blob/master/notebooks/predict_cell_types.ipynb)



## Links
[Project webpage](https://marionilab.github.io/RabbitGastrulation2022/)
[Project github repository](https://github.com/MarioniLab/RabbitGastrulation2022)
[scrabbitpy](https://github.com/dkeitley/scrabbitpy)

